README.md
# elections-carousel

[![Build Status](https://travis-ci.org/hfagerlund/elections-carousel.svg?branch=master)](https://travis-ci.org/hfagerlund/elections-carousel)

Carousel presentation of data from a JSON feed of election results.

Enables scrolling through the election results for all ridings using the Next and Back controls (via mouse or keyboard), displaying the number of votes and automatically calculating the percentage of the vote won by each candidate in each riding.

<img style="max-width:100%;" alt="screenshot of elections-carousel page" src="/screenshots/screenshot_elections-html.png" align="center" /><br />
*Figure 1: screenshot of elections.html page*

Refer to the [documentation](https://hfagerlund.github.io/elections-carousel/) for complete details of customization and usage.
- - -

## Requirements
- npm
  - Babelify
  - Browserify
  - jQuery
  - Istanbul
  - Mustache
  - Tape
  - SASS
- [Yarn](https://github.com/yarnpkg/yarn)

(All copyrights for the above remain with their respective owners.)

- - -

## Quick start

```
# clone the repo
$ git clone https://github.com/hfagerlund/elections-carousel.git

# install dependencies
$ yarn

# launch local server (for data feed)
$ http-server

## browse to http://127.0.0.1:8080/elections.html

```

## Features

* Automatic [bundle hash updates](https://hfagerlund.github.io/elections-carousel/developer-guide/usage/#running-the-build);
* [Customizable](https://github.com/hfagerlund/elections-carousel#parameters);
* Modular (UMD support);
* [Keyboard support](https://hfagerlund.github.io/elections-carousel/developer-guide/usage/#keyboard-support).

## Parameters:
The following carousel features can be (optionally) customized:

| Parameter | Type<br><a id="default" name="default">(Default value) | Description |
| --- | --- | --- |
| **url** | url<br>*(default value: http://127.0.0.1:8080/src/assets/fixtures/results.js)* | The URL for the (JSON) data feed |
| **callback** | `string`<br>*(default value: gNews_getRidingDetailsCallback)* | The JSON-P callback function name |
| **headingtext** | `string`<br>*(default value: 2011 Election Results)* | The text content of the `<h1></h1>` tags |
| **slidewidth** | `int`<br>*(default value: 600)* | The width (in **px**) of each riding 'slide' |
| **templateid** | `string`<br>*(default value: #viewTemplate)* | The template identifier (ie. within the **template file** - refer to [templatepath](https://github.com/hfagerlund/elections-carousel#templatepath)) |
| <a id="templatepath" name="templatepath">**templatepath**</a> | path<br>*(default value: http://127.0.0.1:8080/src/assets/templates/viewFull.mustache)* | The full (absolute) path to the Mustache template file |
| **transitiontime** | `int`<br>*(default value: 300)* | The time span (in **milliseconds**) for carousel slide movement/rotation |

- - -

### Initialization:

#### No arguments:
If no arguments are provided when it is initialized, the module will run using its **default values**.

**Example:**

```javascript
ElectionsCarousel.init({});
```

is equivalent to:

```javascript
ElectionsCarousel.init({
                url            :'http://127.0.0.1:8080/src/assets/fixtures/results.js',
                callback       :'gNews_getRidingDetailsCallback',
                headingtext    : '2011 Election Results',
                slidewidth     : 600,
		templateid     : '#viewTemplate',
		templatepath   : 'http://127.0.0.1:8080/src/assets/templates/viewFull.mustache',
		transitiontime : 300
              });
```

- - -

#### Overriding a selected parameter's default value:
**Example**: Initialize the module with an explicit `'transitiontime'` value of 100 msec in order to override the default (of 300 msec) as follows:

```javascript
ElectionsCarousel.init({transitiontime: 100});
```
(NOTE: [Default values](https://github.com/hfagerlund/elections-carousel#default) will be used for all other unspecified parameters.)

- - -
## Running the Unit Tests

### On the command line:
Run all (Tape) tests, and read the results, on the command line using the following:

```
$ yarn run test

```

- - -
## License
Copyright (c) Heini Fagerlund. Licensed under the [MIT license](http://opensource.org/licenses/mit-license.php).
(See [LICENSE](https://github.com/hfagerlund/elections-carousel/blob/master/LICENSE).)


CHANGELOG.md
# Changelog:

- - -
## Unreleased (2.1.0)
*(last updated: 2018-01-14)*

### New features:
* Toolchain / workflow improvements:
  * Automated code formatting (Prettier);
  * Integrated bundle hash updates (HTML Webpack Plugin);
  * Improved dependency management (Yarn);
   * (Travis) Node version upgrade (from 7 to 8).

- - -
## 2.0.0 - 2017-10-02
#### New features:
* Enabled more parameters to be user-configurable
* Added scripts to enable testing (by automating addition of test code blocks to source)

#### Changed:
* Refactored module methods - improved readability
* Relocated classnames, text content from script to (Mustache) template - improved separation of concerns
* Improved usability, look-and-feel:
  * Repositioned carousel control elements (fixed position, independent of height of slide)
  * More obvious 'highlighting' of winning candidate information
  * Realigned layout of text results/'legend' for graphs - easier readability
* Updated screenshot; added README section on 'Parameters'; added corresponding links to [documentation](https://hfagerlund.github.io/elections-carousel/) for sections relocated from README

#### Known Issues:
* comment-out.sh adds *multiple* leading forward slashes (issue is currently remedied in uncomment.sh which strips out all leading slashes - TODO: fix this in the comment-out.sh script itself).

- - -
## 1.0.1 - 2017-07-08
#### Fixed:
* Accessibility:
  * Replaced carousel control elements in template and modified corresponding attributes, behaviour, styling

- - -
## 1.0.0 - 2017-04-26
#### New features:
* Universal Module Definition (UMD) support
* Mustache client-side templating (cleaner separation of concerns)
* Support for coding style consistency (added .editorconfig)
* Code coverage reporting (using Istanbul)
* Preliminary cleanup for (potential future) npm package creation
* Integrated with Travis CI server
* Added local server setup (for data feed)
* Introduced CHANGELOG (ie. extracted into a separate document from README)

#### Changed:
* Launched [project documentation site](https://hfagerlund.github.io/elections-carousel/)
* Replaced QUnit test harness with TAPE (to support Node as well as browser)
  * ~~Upgraded QUnit version (from v.1.14.0 to v.2.0.1)~~
* Directory structure modifications (eg. relocated sample data feed)
* Modified markup, semantic improvements (eg. 'grouping' of text summary elements with container)
* Accessibility improvements:
  * Added ARIA role and support for carousel control activation using space bar, keyboard
* Styling changes:
  * Revised container identifiers (new classnames)
  * Fixed main heading display, indentation
  * Carousel controls' arrow icons and disabled/enabled state styling relocated from markup and script respectively
  * Removed 'print' stylesheet ('media' attribute for main stylesheet changed from 'screen' to 'all')
* Improved (JavaScript) linting
* Simplified dependency management, augmented build toolchain
  * Added Webpack (v.2.3.3)
  * Build runs using npm (removed Grunt)
  * Dependencies *fully* managed using npm (removed Bower)
* Updated format of README

- - -
## 0.4.0 - 2016-01-13
#### New features: 
* New look-and-feel (GUI improvements):
  * Switched from horizontal to vertical bars
   * Height of bars provides visualization of percentage of votes per candidate
  * Styling improvements: font updates, gradients, party 'labels' (included in list format), emphasized riding vote totals
  * (Minor) text copy changes

#### Changed:
* Improved alphabetical ordering of styling rules
* Removed unnecessary role attributes
* Updated screenshot

- - -
## 0.3.0 - 2016-01-13
#### Changed:
* Results listed in descending numeric order (reordered)

- - -
## 0.2.1 - 2015-12-14
#### Fixed:
* Removed deprecated code

- - -
## 0.2.0 - 2015-12-14
#### New features:
* Defined  global, read-only variables (JSHint strict mode)
* Automated package management

#### Changed:
* Refactored code, performance improvements
* Improved vote percentages (to 2 decimal places; totals 100%)
* Accessibility enhancement: 'Elected' text rendering

#### Fixed:
* Button (enabled/disabled) state bugs

- - -
## 0.1.1 - 2015-07-05 
#### New features:
* Enhanced button states

- - -
## 0.1.0 - 2015-07-05 
#### New features:
* Initial version

- - -
#### *Notes*: 
* All dates shown in YYYY-MM-DD format
* Using [SemVer](http://semver.org/)
* Using GitHub flow


src/js/elections/elections-carousel.js
/* 
 * elections-carousel
 * @copyright (c) Heini Fagerlund
 * @version 2.0.0
 * @license MIT
 * 
 */

//var jQuery = require('jquery'); //TEST_CODE1

require('../../sass/screen.scss');

/*
 * Usage:
 *
 * [a.] All parameters (shown below with default values):
 * ElectionsCarousel.init({
 *   url            :'http://127.0.0.1:8080/src/assets/fixtures/results.js', //data feed URL
 *   callback       :'gNews_getRidingDetailsCallback', //JSON-P callback
 *   headingtext    : '2011 Election Results', //h1 tag text content
 *   slidewidth     : 600, //width of each riding 'slide'
 *   templateid     : '#viewTemplate', //template identifier (within template file)
 *   templatepath   : 'http://127.0.0.1:8080/src/assets/templates/viewFull.mustache', //path to Mustache template file
 *   transitiontime : 300, //milliseconds (for slide movement)
 });
 *
 * [b.] No parameters (uses default values listed above):
 * ElectionsCarousel.init({});
 *
 * [c.] Overriding selected parameter's default value:
 * ElectionsCarousel.init({transitiontime: 100});
 * 
 */
var ElectionsCarousel = (function($) {
  'use strict';

  var _config = {
    css: {
      control: '.ctrl',
      controlBack: '.ctrl--back',
      controlNext: '.ctrl--next',
      disabledClassname: 'ctrl--disabled', //not a typo; do not include dot
      listCandidateRoot: '.riding__list-candidate--',
      mainContainer: '.ec-results',
      ridingIdRoot: '#art_',
      ridingWrapper: '.ec-results__wrapper-ridings'
    }
  };

  var _initialize = {
    initWrapperPosition: 0,
    maxRightHorizontalPosition: 0,
    movementCounter: 0 //control click/key tracking
  };

  var _roundToTwoDec = function(num) {
    /*
     * source: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round
     */
    return +(Math.round(num + 'e+2') + 'e-2');
  };

  var _setBarGraphHeights = function(options) {
    /*
     * set heights of bars on graph
     */
    var mainStyle = document.styleSheets[0];
    var oLength = mainStyle.cssRules.length;
    var barHeight = 250 * (options.percentage / 100);
    mainStyle.insertRule(
      _config.css.ridingIdRoot +
        options.id +
        ' ' +
        _config.css.listCandidateRoot +
        options.party +
        '{height:' +
        barHeight +
        'px!important;position:relative}',
      oLength
    );
  };

  var _calculateCandidateVotePercentages = function(data) {
    /*
     * calculate candidate vote percentages
     */
    for (var i = 0; i < data.length; i++) {
      var riding = data[i];
      var ridingId = riding.id;
      var ridingResults = riding.results;
      var ridingResultsVotes = ridingResults.map(function(item) {
        return item.votes;
      });
      //total votes for riding
      var totalRidingResultsVotes = ridingResultsVotes.reduce(function(pv, cv) {
        return pv + cv;
      }, 0);
      riding.voteTotal = totalRidingResultsVotes;

      riding.results.forEach(function(candidate) {
        var candidateParty = candidate.partyCode.toLowerCase();
        var candidateVotes = candidate.votes;
        var candidateVotePercentage = _roundToTwoDec(
          candidateVotes / totalRidingResultsVotes * 100
        );
        candidate.votePercentage = candidateVotePercentage;
        _setBarGraphHeights({
          id: ridingId,
          party: candidateParty,
          percentage: candidateVotePercentage
        });
        //winning candidate
        candidate.winner = candidate.ifWon = candidate.ifElected = candidate.isElected;
      });
    }
  };

  var _sortCandidatesByDescendingVotes = function(data) {
    for (var i = 0; i < data.length; i++) {
      var riding = data[i];
      var ridingResults = riding.results;
      //descending order
      var ridingResultsByVotes = ridingResults.sort(function(a, b) {
        return b.votes - a.votes;
      });
      $(this).parent.resultsByVotes = ridingResultsByVotes;
    }
  };

  var _rotateSlides = function(newWrapperPosition, transitionTime) {
    $(_config.css.ridingWrapper).animate({ left: newWrapperPosition }, transitionTime, function() {
      /* action on animation completion */
    });
  };

  var _disableCtrl = function(ctrlClass) {
    var ctrl = $(ctrlClass);
    ctrl.addClass(_config.css.disabledClassname);
    ctrl.attr('aria-disabled', 'true'); //not strictly required anymore (since button element natively supports 'disabled' attribute)
    ctrl.attr('disabled', 'disabled');
  };

  var _reEnableCtrl = function(ctrlClass) {
    var ctrl = $(ctrlClass);
    ctrl.removeClass(_config.css.disabledClassname);
    ctrl.attr('aria-disabled', 'false'); //redundant
    ctrl.removeAttr('disabled');
  };

  var _calculateHorizontalPosition = function(whichControl, data) {
    var slideWidth = data.slideWidth + 4; //additional (4px) for borders
    var ridingCounter = 0;
    data.forEach(function(riding) {
      ridingCounter++;
    });
    var numberOfSlides = ridingCounter;

    var wrapperWidth = numberOfSlides * slideWidth; //total (horizontal) width of all slides
    //limits for horizontal positioning of wrapper
    var maxLeftHorizontalPosition = -((numberOfSlides - 1) * slideWidth); //equivalent to: -(wrapperWidth - (1*slideWidth));

    if (whichControl.indexOf('back') != -1) {
      _initialize.movementCounter--;
    } else if (whichControl.indexOf('next') != -1) {
      _initialize.movementCounter++;
    }

    var newWrapperPosition =
      _initialize.initWrapperPosition - _initialize.movementCounter * slideWidth;

    if (newWrapperPosition >= _initialize.maxRightHorizontalPosition) {
      _rotateSlides(newWrapperPosition, data.transitionTime);
      //disable/re-enable controls
      _disableCtrl(_config.css.controlBack);
      _reEnableCtrl(_config.css.controlNext);
      if (newWrapperPosition > _initialize.maxRightHorizontalPosition) {
        //reached 1st slide/max. Back limit
        //stop movement and reset for next click
        _initialize.movementCounter = 0;
        newWrapperPosition = _initialize.maxRightHorizontalPosition;
      }
    } else if (newWrapperPosition <= maxLeftHorizontalPosition) {
      _rotateSlides(newWrapperPosition, data.transitionTime);
      //disable/re-enable controls
      _disableCtrl(_config.css.controlNext);
      _reEnableCtrl(_config.css.controlBack);
      if (newWrapperPosition < maxLeftHorizontalPosition) {
        //reached last slide - ie. max Next limit
        //stop movement and reset for next click
        _initialize.movementCounter = numberOfSlides - 1;
        newWrapperPosition = maxLeftHorizontalPosition;
      }
    } else {
      //within allowable range of movement
      _rotateSlides(newWrapperPosition, data.transitionTime);
      _reEnableCtrl(_config.css.controlNext);
      _reEnableCtrl(_config.css.controlBack);
    }
  };

  var _bindCtrlListener = function(events, data) {
    //jQuery 1.7+
    $(document).on(events, _config.css.control, function(e) {
      if (e.type === 'click') {
        //Enter key, spacebar, or mouse click
        var whichControl = $(e.target).attr('data-js');
        _calculateHorizontalPosition(whichControl, data);
      }
    });
  };

  var _renderTemplate = function(data) {
    _sortCandidatesByDescendingVotes(data);
    _calculateCandidateVotePercentages(data);

    //attach Mustache template
    $.get(data.templatePath, function(templates) {
      var template = $(templates)
        .filter(data.templateId)
        .html();
      $(_config.css.mainContainer).append(Mustache.render(template, data));
    });

    data.lower = function() {
      return function(text, render) {
        return render(text).toLowerCase();
      };
    };

    //attach behavior to the controls' onkeydown and onclick events
    _bindCtrlListener('keydown click', data);
  };

  var extendData = function(
    data,
    headingTxt,
    slidewidth,
    templateid,
    templatepath,
    transitiontime
  ) {
    var userDefinedSettings = {
      headingText: headingTxt,
      slideWidth: slidewidth,
      templateId: templateid,
      templatePath: templatepath,
      transitionTime: transitiontime
    };
    $.extend(data, userDefinedSettings);
    //      return data.pollsReported; //TEST_CODE2
    _renderTemplate(data);
  };

  var _getJsonData = function(
    url,
    callback,
    headingtext,
    slidewidth,
    templateid,
    templatepath,
    transitiontime
  ) {
    $.ajax({
      type: 'GET',
      url: url,
      contentType: 'application/json; charset=UTF-8',
      jsonpCallback: callback,
      dataType: 'jsonp',
      custom: headingtext
    }).done(function(data, custom) {
      var headingTxt = this.custom;
      extendData(data, headingTxt, slidewidth, templateid, templatepath, transitiontime);
    });
  };

  var main = function(options) {
    //set default values
    options.url =
      typeof options.url !== 'undefined'
        ? options.url
        : 'http://127.0.0.1:8080/src/assets/fixtures/results.js';
    options.callback =
      typeof options.callback !== 'undefined' ? options.callback : 'gNews_getRidingDetailsCallback';
    options.headingtext =
      typeof options.headingtext !== 'undefined' ? options.headingtext : '2011 Election Results';
    options.slidewidth = typeof options.slidewidth !== 'undefined' ? options.slidewidth : 600;
    options.templateid =
      typeof options.templateid !== 'undefined' ? options.templateid : '#viewTemplate';
    options.templatepath =
      typeof options.templatepath !== 'undefined'
        ? options.templatepath
        : 'http://127.0.0.1:8080/src/assets/templates/viewFull.mustache';
    options.transitiontime =
      typeof options.transitiontime !== 'undefined' ? options.transitiontime : 300;

    _getJsonData(
      options.url,
      options.callback,
      options.headingtext,
      options.slidewidth,
      options.templateid,
      options.templatepath,
      options.transitiontime
    );
  };

  return {
    init: main
    //    , extendData: extendData //TEST_CODE2
  };
})(jQuery);

module.exports = ElectionsCarousel;
